### PR TITLE
fix: retry applying lxd profiles

### PR DESCRIPTION
**Pull request by SimonRichardson**
This ensures that if the provider profiles don't match the expected profiles, then return false from the verification process. We can shortcut the whole process by checking if the lengths match before doing any additional computation. Now that we know the sets are the same length, we can use the set difference method to check if both contain the same profiles. We couldn't use the set difference method previously as it was biased to the left side, so any additional items in the right side would be ignored.

Changing the logic gives us more reliable verification of the charm profiles, with the additional tests baking in that the profiles might be different.

Additionally, if the profiles reported back from the broker are already the profiles requested by the charms/model, we reflect this back to the machine entity. The reason for this is that sometimes setting the profiles via the broker can sometimes cause it to actually apply the profiles but report back an error. In this case we can see the profiles when reading from the broker, we can see they match what we expect, but we never reflect this back to the machine entity, leading to units with a charm that depends on that profile being applied to the machine, unable to proceed with startup.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Ensure that the units come up correctly.

```sh
$ juju bootstrap lxd test
$ juju deploy microk8s --channel 1.28/stable --config rbac=true --constraints "virt-type=virtual-machine" -n 3
```
 
## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2072769

**Jira card:** [JUJU-6384
](https://warthogs.atlassian.net/browse/JUJU-6384)

### CODE REVIEW

File: worker/instancemutater/mutater.go [hunk 1]
```diff
@@ -237,13 +237,13 @@ func (m MutaterMachine) processMachineProfileChanges(info *instancemutater.UnitP
...
 
-	verified, err := m.verifyCurrentProfiles(string(info.InstanceId), expectedProfiles)
+	verified, currentProfiles, err := m.verifyCurrentProfiles(string(info.InstanceId), expectedProfiles)
 	if err != nil {
 		return report(errors.Annotatef(err, "%s", m.id))
 	}
 	if verified {
 		m.logger.Infof("no changes necessary to machine-%s lxd profiles (%v)", m.id, expectedProfiles)
-		return report(nil)
+		return report(m.machineApi.SetCharmProfiles(currentProfiles))
```
Comment (hmlanigan) [inline worker/instancemutater/mutater.go:246]: This is changing the behavior juju as the default and juju-model-name profiles are now being written to instanceData where they were not before. We've only been writing the charm profiles, applied, not all profiles. We're also writing profiles to every lxd machine, not just ones with charm specified profiles.

instanceData from the lxd controller:
```
	"machineid" : "0",
	"charm-profiles" : [
		"default",
		"juju-controller"
	]
```

**Review by hmlanigan — CHANGES_REQUESTED** [review]
My local machines do not reproduce this issue, a bit difficult to validate QA. However I did find a problem.

File: worker/instancemutater/mutater.go (hunk 1 above)
Comment (hpidcock) [reply worker/instancemutater/mutater.go:246]: > We've only been writing the charm profiles, applied, not all profiles.
verifyCurrentProfiles now calls LXDProfileNames, the value is reflected back to instanceData.
AssignLXDProfiles currently calls LXDProfileNames, the value is reflected back to instanceData.
I don't see the change in behaviour here.

> We're also writing profiles to every lxd machine, not just ones with charm specified profiles.
By this you mean we are reflecting in the instanceData for every machine the profiles that are applied?

**Comment by hpidcock** [comment]
> My local machines do not reproduce this issue, a bit difficult to validate QA. However I did find a problem.

I'm not sure what changes you are requesting here. I'll leave this to @SimonRichardson to chat with you about.

**Comment by SimonRichardson** [comment]
@hmlanigan we're not recording any data that isn't already being recorded. The first commit is just triggering when that could occur. As the set methods are always left-biased, there was a small edge case where we wouldn't trigger in every scenario.

The second commit always records the profile names to prevent any deviation. The `AssignLXDProfiles`, returns the same information as `verifyCurrentProfiles` (see line 436 of the code below).

https://github.com/juju/juju/blob/69abb5baadd616b5cf287dad3ef21524e58d0265/container/lxd/manager.go#L426-L467

There is no additional information that we're not already recording.

To summarise:

 1. This PR solves a case that we're not always correctly verified.
 2. We record the information anyway to ensure that we're eventually consistent.

**Comment by hmlanigan** [comment]
Are you advocating that "charm" profiles written to state and charm profiles displayed in juju status should be machine profiles? I disagree, however if we must, we should be consistent. The provisioner writing charm profiles to state and the instance mutater writing both the default, juju and charm profiles to state. 

We wrote a method to remove default and the model profile from the list: https://github.com/juju/juju/blob/5e245caf05620ddbfd96f7c8f85056368852e0f0/core/lxdprofile/name.go#L33 

I checked `juju status --format` and it still displayed correctly, weeding out the `default` and `juju-<model>` profiles from the machine. Only instanceData.charmProfiles will be misnamed.

File: worker/instancemutater/mutater.go (hunk 1 above)
Comment (hmlanigan) [reply worker/instancemutater/mutater.go:246]: @hpidcock yes - we used to reflect only charm specific profiles in the instanceData. Now all profiles are being reflected in the instanceData.

File: state/machine.go [hunk 2]
```diff
@@ -430,29 +430,33 @@ func (m *Machine) CharmProfiles() ([]string, error) {
...
+		currentProfiles, err := m.CharmProfiles()
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		mProfilesSet := set.NewStrings(mProfiles...)
-		if mProfilesSet.Union(set.NewStrings(profiles...)).Size() == mProfilesSet.Size() {
+		currentProfileSet := set.NewStrings(currentProfiles...)
+		newProfileSet := set.NewStrings(profiles...)
+		if currentProfileSet.Size() == newProfileSet.Size() &&
+			currentProfileSet.Intersection(newProfileSet).Size() == currentProfileSet.Size() {
```
Comment (SimonRichardson) [inline state/machine.go:448]: **nit:** use difference and then you can check it against 0

**Review by SimonRichardson — COMMENTED** [review]
LGTM

**Review by hmlanigan — APPROVED** [review]
LGTM, Thanks @hpidcock

File: state/machine.go (hunk 2 above)
Comment (hpidcock) [reply state/machine.go:448]: Difference is an exclusive left outer join. It doesn't account for the values in the right that isn't in the left set.

Comment (hpidcock) [reply state/machine.go:448]: This equality check exploits: If the size of the inner join set is equal to the size of both the left set and the right set, the left and right must be equal.

Comment (SimonRichardson) [reply state/machine.go:448]: If the size equality is checked (which it is), then an intersection can be dropped in favour of a difference and checking if it's nonzero.
